### ISSUE TITLE
Permissions: Template/Scaffolder:  READ missing User object in handle call, CREATE and DELETE have it

### ISSUE THREAD

**Issue by FilipSwiatczak** [backstage/backstage#12390]
## Expected Behavior
Template Access control:
- permission.ts has hardcoded "admin" usernames that return ALLOW for all requests instantly when identity.userEntityRef matches
- non-Template entities return ALLOW for READ but check for Ownership for DELETE (identity.ownershipEntityRefs)
- Template entities check Ownership OR "unrestricted" label (catalogConditions.hasLabel)
- Under the hood: Read permissions triggered when opening Create page/scaffolder, the handle function should receive existing BackstageIdentityResponse aka. _user for all READ attempts.

## Actual Behavior
**Production**:
- READ permission requests for Create/Scaffolder have "undefined" _user passed in handle function call, making admin or ownership checks impossible.
- Delete restriction works correctly in catalog - _user is passed and non-owners are denied Deletion, while Admins are also recognised
- Create permission requests also have the _user passed in

**Local development** - everything works as expected

## Steps to Reproduce
Our packages/backend/src/plugins/permission.ts
```typescript
const isAdmin = (user: BackstageIdentityResponse | undefined) =>
  ['filipswiatczak', '...', '...'].some(admin =>
    user?.identity.userEntityRef.includes(admin),
  );

class CustomPolicy implements PermissionPolicy {
  async handle(
    request: PolicyQuery,
    _user?: BackstageIdentityResponse,
  ): Promise<PolicyDecision> {
    console.log(
      `User: identity.type - ${_user?.identity.type} User: identity.userEntityRef - ${_user?.identity.userEntityRef} User: identity.ownershipEntityRefs.length - ${_user?.identity.ownershipEntityRefs.length} Request: type - ${request.permission.type}; name - ${request.permission.name}; action - ${request.permission.attributes.action}`,
    );
    // exempting admins from permission checks
    if (!_user?.identity.userEntityRef || isAdmin(_user)) {
      return { result: AuthorizeResult.ALLOW };
    }
    // Read permissions
    if (isPermission(request.permission, catalogEntityReadPermission)) {
      return createCatalogConditionalDecision(request.permission, {
        anyOf: [
          catalogConditions.isEntityKind([
            'Domain',
            'Component',
            'System',
            'API',
            'Group',
            'User',
            'Resource',
            'Location',
          ]),
          // Templates must be owned by the user, his team, or parent group to be visible
          {
            allOf: [
              catalogConditions.isEntityKind(['Template']),
              {
                anyOf: [
                  catalogConditions.isEntityOwner(
                    _user?.identity.ownershipEntityRefs ?? [],
                  ),
                  catalogConditions.hasLabel('unrestricted'),
                ],
              },
            ],
          },
        ],
      });
    }
    // deletion check for entities owned by user
    if (
      isDeletePermission(request.permission) &&
      // Narrow type of `request.permission` to `ResourcePermission<'catalog-entity'>
      isResourcePermission(request.permission, RESOURCE_TYPE_CATALOG_ENTITY)
    ) {
      return createCatalogConditionalDecision(
        request.permission,
        catalogConditions.isEntityOwner(
          _user?.identity.ownershipEntityRefs ?? [],
        ),
      );
    }
    return { result: AuthorizeResult.ALLOW };
  }
}

export default async function createPlugin(
  env: PluginEnvironment,
): Promise<Router> {
  return await createRouter({
    config: env.config,
    logger: env.logger,
    discovery: env.discovery,
    policy: new DeletionPolicy(),
    identity: IdentityClient.create({
      discovery: env.discovery,
      issuer: await env.discovery.getExternalBaseUrl('auth'),
    }),
  });
}
```

## Context
We are attempting to control Template access to owner groups, with exception to admin core group

Authenticated user below, also admin, has entity claims present assigned by BB Cloud resolver:
![image](https://user-images.githubusercontent.com/30108240/176897321-2b5ac5a7-5ca6-4307-8e77-b29fcaa995d4.png)
Production Graylog printing request and _user details for permission attempts. This is for user authenticated with Bitbucket Cloud, with assigned entity claims. **Note Create and Delete permissions have the _user object, the READ has "undefined"**
![image](https://user-images.githubusercontent.com/30108240/176894114-10f0fba7-d0b0-484e-b6ec-5a2244fa1581.png)
![image](https://user-images.githubusercontent.com/30108240/176894234-da056692-5a63-4c7c-a905-feb853cba5e2.png)
Working in local:
![image](https://user-images.githubusercontent.com/30108240/176904524-80ef557f-aac0-45bf-8def-3a558a2b81ac.png)

## Your Environment
Bitbucket Cloud resolves Identity by UUID:
```typescript
providerFactories: {
      bitbucket: providers.bitbucket.create({
        signIn: {
          resolver:
            providers.bitbucket.resolvers.userIdMatchingUserEntityAnnotation(),
        },
      }),
    },
```
- app-config.production.yaml backend section:
```yaml
backend:
  baseUrl: ${BACKSTAGE_URL}:${BACKSTAGE_BACKEND_PORT}
  listen:
    port: ${BACKSTAGE_BACKEND_PORT}
```
- app-config.yaml backend section, this works fine in local dockerised tests:
```yaml
permission:
  enabled: true
backend:
  baseUrl: ${BACKSTAGE_URL}:${BACKSTAGE_BACKEND_PORT}
  listen:
    port: ${BACKSTAGE_BACKEND_PORT}
  auth:
    keys:
      - secret: ${AUTH_BITBUCKET_CLIENT_SECRET}  
```
- app-config.local.yaml backend section, this works fine:
```yaml
backend:
  baseUrl: http://localhost:7007
  listen:
    port: 7007
```
- Output of `yarn backstage-cli info`: 

```text
local - **working fine**:
OS:   Darwin 21.2.0 - darwin/x64
node: v14.17.4
yarn: 1.22.17
cli:  0.17.2 (installed)

prod - **failing** as described:
OS:   Linux 5.10.47-linuxkit - linux/x64
node: v16.15.1
yarn: 1.22.19
cli:  0.17.2 (installed)

Dependencies, same in all envs:
  @backstage/app-defaults                                  1.0.3
  @backstage/backend-common                                0.14.0
  @backstage/backend-tasks                                 0.3.2
  @backstage/backend-test-utils                            0.1.25
  @backstage/catalog-client                                1.0.3
  @backstage/catalog-model                                 1.0.3
  @backstage/cli-common                                    0.1.9
  @backstage/cli                                           0.17.2
  @backstage/config-loader                                 1.1.2
  @backstage/config                                        1.0.1
  @backstage/core-app-api                                  1.0.3
  @backstage/core-components                               0.9.5
  @backstage/core-plugin-api                               1.0.3
  @backstage/dev-utils                                     1.0.3
  @backstage/errors                                        1.0.0
  @backstage/integration-react                             1.1.1
  @backstage/integration                                   1.2.1
  @backstage/plugin-analytics-module-ga                    0.1.17
  @backstage/plugin-api-docs                               0.8.6
  @backstage/plugin-app-backend                            0.3.33
  @backstage/plugin-auth-backend                           0.14.1
  @backstage/plugin-auth-node                              0.2.2
  @backstage/plugin-badges-backend                         0.1.27
  @backstage/plugin-badges                                 0.2.30
  @backstage/plugin-bazaar-backend                         0.1.17
  @backstage/plugin-bazaar                                 0.1.21
  @backstage/plugin-bitbucket-cloud-common                 0.1.0
  @backstage/plugin-catalog-backend-module-bitbucket-cloud 0.1.0
  @backstage/plugin-catalog-backend                        1.2.0
  @backstage/plugin-catalog-common                         1.0.3
  @backstage/plugin-catalog-graph                          0.2.18
  @backstage/plugin-catalog-import                         0.8.9
  @backstage/plugin-catalog-react                          1.1.1
  @backstage/plugin-catalog                                1.3.0
  @backstage/plugin-explore-react                          0.0.18
  @backstage/plugin-explore                                0.3.37
  @backstage/plugin-github-actions                         0.5.6
  @backstage/plugin-home                                   0.4.22
  @backstage/plugin-jenkins-backend                        0.1.23
  @backstage/plugin-jenkins-common                         0.1.5
  @backstage/plugin-jenkins                                0.7.5
  @backstage/plugin-lighthouse                             0.3.6
  @backstage/plugin-org                                    0.5.6
  @backstage/plugin-permission-backend                     0.5.8
  @backstage/plugin-permission-common                      0.6.2
  @backstage/plugin-permission-node                        0.6.2
  @backstage/plugin-permission-react                       0.4.2
  @backstage/plugin-proxy-backend                          0.2.27
  @backstage/plugin-scaffolder-backend                     1.3.0
  @backstage/plugin-scaffolder-common                      1.1.1
  @backstage/plugin-scaffolder                             1.3.0
  @backstage/plugin-search-backend-module-elasticsearch    0.1.5
  @backstage/plugin-search-backend-module-pg               0.3.4
  @backstage/plugin-search-backend-node                    0.6.2
  @backstage/plugin-search-backend                         0.5.3
  @backstage/plugin-search-common                          0.3.5
  @backstage/plugin-search-react                           0.2.1
  @backstage/plugin-search                                 0.9.0
  @backstage/plugin-sentry                                 0.3.44
  @backstage/plugin-sonarqube                              0.3.6
  @backstage/plugin-stack-overflow                         0.1.2
  @backstage/plugin-tech-radar                             0.5.13
  @backstage/plugin-techdocs-backend                       1.1.2
  @backstage/plugin-techdocs-module-addons-contrib         1.0.1
  @backstage/plugin-techdocs-node                          1.1.2
  @backstage/plugin-techdocs-react                         1.0.1
  @backstage/plugin-techdocs                               1.2.0
  @backstage/plugin-user-settings                          0.4.5
  @backstage/release-manifests                             0.0.4
  @backstage/test-utils                                    1.1.1
  @backstage/theme                                         0.2.15
  @backstage/types                                         1.0.0
  @backstage/version-bridge                                1.0.1
```

**Comment by freben** [comment]
@Joonpark13 @vinzscam This one might be for you! Could this be a case of a permission that should have been a resource permission but isn't?

**Comment by vinzscam** [comment]
Hi @FilipSwiatczak,
when are you experiencing the issue? Is it when you are actually viewing the `/create` page?

**Comment by FilipSwiatczak** [comment]
hello @vinzscam, thanks for looking into this. Yes it's when viewing the `/create` page. In local I can see the Read permission requests having the _user passed in while in production the _user is undefined.

**Comment by regicsolutions** [comment]
@FilipSwiatczak wondering if you were ever able to get passed this issue?

**Comment by padupe** [comment]
This implementation interests me too. Are there any updates?

**Comment by freben** [comment]
I don't think it was ever addressed right? Any other takers for implementing it?

**Comment by cpgonz2** [comment]
could you please share also the types.ts?

**Comment by kevinstafford** [comment]
We ran into this issue, and it turned out to be related to some customizations we made to `apis.ts`. We added a call to createApiFactory for the catalogApiRef api, and when we deleted that change the issue went away. I'm not entirely sure what was wrong with that bit of code, but it turns out what we had intended to do was no longer necessary, so we just reverted our changes. Hopefully this is helpful to someone encountering the same issue.
